docs/reference/query-dsl/script-score-query.asciidoc

```diff
@@ -0,0 +1,263 @@
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "decayGeoExp(params.origin, params.scale, params.offset, params.decay, doc['location'].value)",
+    "params": {
+        "origin": "40, -70.12",
+        "scale": "200km",
+        "offset": "0km",
+        "decay" : 0.2
+    }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+
+===== Decay functions for date fields
+
+* `double decayDateLinear(String originStr, String scaleStr, String offsetStr, double decay, JodaCompatibleZonedDateTime docValueDate)`
+
+* `double decayDateExp(String originStr, String scaleStr, String offsetStr, double decay, JodaCompatibleZonedDateTime docValueDate)`
+
+* `double decayDateGauss(String originStr, String scaleStr, String offsetStr, double decay, JodaCompatibleZonedDateTime docValueDate)`
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "decayDateGauss(params.origin, params.scale, params.offset, params.decay, doc['date'].value)",
+    "params": {
+        "origin": "2008-01-01T01:00:00Z",
+        "scale": "1h",
+        "offset" : "0",
+        "decay" : 0.5
+    }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+NOTE: Decay functions on dates are limited to dates in the default format
+and default time zone. Also calculations with `now` are not supported.
+
+
+==== Faster alternatives
+Script Score Query calculates the score for every hit (matching document).
+There are faster alternative query types that can efficiently skip
+non-competitive hits:
+
+* If you want to boost documents on some static fields, use
+ <<query-dsl-feature-query, Feature Query>>.
+
+
+==== Transition from Function Score Query
+We are deprecating <<query-dsl-function-score-query, Function Score>>, and
+Script Score Query will be a substitute for it.
+
+Here we describe how Function Score Query's functions can be
+equivalently implemented in Script Score Query:
+
+===== `script_score`
+What you used in `script_score` of the Function Score query, you
+can copy into the Script Score query. No changes here.
+
+===== `weight`
+`weight` function can be implemented in the Script Score query through
+the following script:
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "params.weight * _score",
+    "params": {
+        "weight": 2
+    }
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+===== `random_score`
+For easy generating not reproducible random values:
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "Math.random()"
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+or set your own reproducible seed using this script:
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "Random rnd = new Random(); rnd.setSeed(doc['field'].value); rnd.nextFloat()"
```

Comment:
I think this should use a seed from a parameter rather than from a document?

Also, I assume this uses the parameter caching so the instance of Random is created once and is kept the same through the execution of the script on each doc? If not, then maybe we should do this as it will make searches with the same seed reproducible?

Reply:
Actually I guess this can't do the caching since we are not using our own function. I think we should implement our own random function that caches the instance of `Random` with the provided seed?